Top-level networks and volumes

michaelsauter opened the pull request.

michaelsauter commented:
@bjaglin Here's a first working version. What do you think?

Two notes:
* The implementation is a bit repetitive, possibly due to my limited knowledge of Go. On the other hand, I think it's easy to understand so I'm relatively fine with it
* Currently there are no options for networks and volumes. That makes the configuration a bit awkard:
```yaml
networks:
  bar: {}
```

bjaglin commented on crane/network.go:
```diff
@@ -0,0 +1,33 @@
+	RawName string
+}
+
+func (n *network) Name() string {
```
should the prefix flag/env variable also apply as namespace for networks/volumes? I'd be tempted to say yes

michaelsauter replied:
Sounds like it should be that way, yes. Unfortunately, implementation is not straight-forward. So maybe we can merge this without it and make that change later?

bjaglin commented on crane/container.go:
```diff
@@ -504,15 +519,25 @@ func (r RunParameters) Volume() []string {
+	volumes := r.Volume()
+	var volumeSources []string
+	for _, volume := range volumes {
+		parts := strings.Split(volume, ":")
```
maybe name `parts[0]` as `source`?

bjaglin commented on crane/config.go:
```diff
@@ -229,6 +268,25 @@ func (c *config) initialize() {
+func (c *config) setNetworkMap() {
+	c.networkMap = make(map[string]Network)
+	for rawName, network := range c.RawNetworks {
+		network.RawName = rawName
```
I guess better nil handling here would allow
```
networks:
  foo:
```

to be valid without the `{}` to force the allocation of an object when unmarshalling

michaelsauter replied:
Ah right, implemented in cb8c09b38b8981609a8b99c0df0d006d5e3438db.

bjaglin commented on crane/config.go:
```diff
@@ -31,7 +36,11 @@ type config struct {
 	RawContainerMap map[string]*container `json:"containers" yaml:"containers"`
 	RawGroups       map[string][]string   `json:"groups" yaml:"groups"`
 	RawHooksMap     map[string]hooks      `json:"hooks" yaml:"hooks"`
+	RawNetworks     map[string]*network   `json:"networks" yaml:"networks"`
```
`RawNetworkMap` ?

michaelsauter replied:
You're right. I decided to standardize them according to the fields in the config: e66b169cd6e2bcd90c175136f85c3e12876899dc

bjaglin commented on crane/container.go:
```diff
@@ -51,16 +51,17 @@ type ContainerInfo interface {
+	RawName     string
+	RawUnique   bool            `json:"unique" yaml:"unique"`
+	RawImage    string          `json:"image" yaml:"image"`
+	RawRequires []string        `json:"requires" yaml:"requires"`
```
did you consider having this somewhere else? I can't find of a better place, but it does feel strange there as most of the other values are first-class Docker citizens

michaelsauter replied:
I think it's actually fine here, `unique` is already a non-Docker value. So is `image`, if you think about it - it sort of belongs into the run config, but that has no option `image`. So I would say non-Docker keys at the container level are okay - as there is no equivalent "command" on the Docker side.

bjaglin commented on crane/unit_of_work.go:
```diff
@@ -289,3 +293,55 @@ func (uow *UnitOfWork) ensureInRequireStarted(name string) {
+}
+
+func (uow *UnitOfWork) prepareNetworks() {
+	for _, n := range uow.requiredNetworks() {
```
Maybe these "required" slices should be exposed so that the debug "Command will be applied to" can reference them as well?

michaelsauter replied:
Good point, added in f27b9ae.

bjaglin commented:
Looks like a solid base! About your 2 notes:

* I think it's just fine the way it is, it's verbose but readable
* If the empty `{}` can be removed as discussed in https://github.com/michaelsauter/crane/pull/248#discussion_r48208082, I think it's fine... I guess the feature can be labeled as experimental so we can always break the format in case it doesn't work out for people in practive.

michaelsauter commented:
Thanks for the great feedback! I addressed most of your comments.

The only major remaining thing is the prefix ...

And yes, integration tests would be great and are desperately needed ... but this PR won't introduce them ;)

bjaglin commented:
:+1:

bjaglin commented:
Looks like Compose has the same model in https://github.com/docker/compose/pull/2564. Maybe some documentation and this could be shipped as experimental?

michaelsauter commented:
@bjaglin Haha, just in time. I added prefix support, an entry to the changelog and some information in the readme. What do you think?

bjaglin commented on README.md:
```diff
@@ -273,6 +276,49 @@ feature is experimental, which means it can be changed or even removed in every
 
 
+### Networking
+Docker networks are support via the top-level config `networks`, which does not
```
`supported`

bjaglin commented on README.md:
````diff
@@ -273,6 +276,49 @@ feature is experimental, which means it can be changed or even removed in every
+  qux:
+```
+
+This feature is experimental, which means it can be changed or even removed in
````
line-wrapped on purpose?

bjaglin commented on README.md:
```diff
@@ -273,6 +276,49 @@ feature is experimental, which means it can be changed or even removed in every
+
+
+### Volumes
+Docker volumes are support via the top-level config `volumes`, which does not
```
`supported`

bjaglin commented:
LGTM apart from the doc cosmetic comments
